src/main/java/net/neoforged/neoforge/common/util/AttributeUtil.java

```diff
@@ -0,0 +1,356 @@
+
+            for (Pair<Holder<Attribute>, AttributeModifier> pair : list) {
+                tooltips.accept(pair.getFirst().value().toComponent(pair.getSecond(), AttributeUtil.getTooltipFlag()));
+            }
+        }
+    }
+
+    /**
+     * Creates a sorted {@link TreeMultimap} used to ensure a stable iteration order of item attribute modifiers.
+     */
+    public static Multimap<Holder<Attribute>, AttributeModifier> sortedMap() {
+        return TreeMultimap.create(Comparator.comparing(Holder::value, Comparator.comparing(BuiltInRegistries.ATTRIBUTE::getKey, ResourceLocation::compareTo)), ATTRIBUTE_MODIFIER_COMPARATOR);
+    }
+
+    /**
+     * Returns a sorted, mutable {@link Multimap} containing all the attribute modifiers on an item stack for the given group.
+     * <p>
+     * This includes attribute modifiers from components (or default modifiers, if not present), enchantments, and the {@link ItemAttributeModifierEvent}.
+     * 
+     * @param stack The stack to query modifiers for.
+     * @param slot  The slot group to query modifiers for.
+     */
+    public static Multimap<Holder<Attribute>, AttributeModifier> getSortedModifiers(ItemStack stack, EquipmentSlotGroup slot) {
+        Multimap<Holder<Attribute>, AttributeModifier> map = AttributeUtil.sortedMap();
+        stack.forEachModifier(slot, (attr, modif) -> {
+            if (attr != null && modif != null) {
+                map.put(attr, modif);
+            } else {
+                LOGGER.debug("Detected broken attribute modifier entry on item {}.  Attr={}, Modif={}", stack, attr, modif);
+            }
+        });
+        return map;
+    }
+
+    /**
+     * Creates a mutable component starting with the char used to represent a drop-down list.
+     */
+    private static MutableComponent list() {
+        return Component.literal(" \u2507 ").withStyle(ChatFormatting.GRAY);
+    }
+
+    /**
+     * Gets the current global tooltip flag. Used by {@link AttributeUtil#addPotionTooltip} since one isn't available locally.
+     * 
+     * @return If called on the client, the current tooltip flag, otherwise {@link TooltipFlag#NORMAL}
+     */
+    private static TooltipFlag getTooltipFlag() {
+        if (FMLEnvironment.dist.isClient()) {
+            return ClientAccess.getTooltipFlag();
+        }
+        return TooltipFlag.NORMAL;
+    }
+
+    /**
+     * Stores a single base modifier (determined by {@link IAttributeExtension#getBaseId()}) and any other children non-base modifiers for the same attribute.
+     * <p>
+     * Used during attribute merging logic within {@link AttributeUtil#applyTextFor}.
+     */
+    private static record BaseModifier(AttributeModifier base, List<AttributeModifier> children) {}
+
+    /**
+     * Client bouncer class to avoid class loading issues. Access to this class still needs a dist check.
+     */
+    private static class ClientAccess {
+        @SuppressWarnings("resource")
```

Comment:
My IDE is telling me this is a redundant suppression. Is this needed for eclipse or was it due to code that you had while prototyping that no longer exists?

Reply:
This is an eclipse thing. Without it, eclipse shows `Resource leak: '<unassigned Closeable value>' is never closed` since `Minecraft` implements `AutoCloseable`

Honestly, I can drop this and just change my eclipse settings to get it to shut up. It does this for every method which returns a `Level` too.